### DISCUSSION TITLE
Nested routes with multiple service

### DISCUSSION THREAD

**Discussion by L4ngu0r** [tokio-rs/axum#742] [Q&A] [answered]
Hello,

I have an API like this :

```
/api/v1.0/area/a
/api/v1.0/area/b
/api/v1.0/area/c
```
`a` & `b` are REST API and `c` are Websockets. I want to use something like `/*key` to cover all REST routes dynamically and provide juste one different handler for websockets, but `Router` doesn't let me to do that as `/*key` already cover the routes.

I tried this approach :

```rust
let rest_api = Router::new().route("/", get(rest_handler).post(rest_handler);
let ws_api = Router::new().route("/", get(ws_handler));

let routes = Router::new()
  .nest("/a", rest_api)
  .nest("/b", rest_api_clone)
  .nest("/c", ws_api);

let app = Router::new().nest("/api/v1.0/area/", routes);
```

It works but it's not very dynamic neither convenient as I have a dozen other REST routes, and I don't want to update this part for every routes.
I tried also to instantiate `Router::new()` through a Vector but struggle to do so.

Can someone suggests me another way ? or maybe I'm missing something ?

Thanks!

**Comment by davidpdrsn** [comment, marked as answer]
You either have to do what you're doing now or capture the path param (with `/*foo`) and match on `foo` and call the respective handler.

**Comment by L4ngu0r** [comment]
Yes thanks, I use match on Path but struggle to find a common type between `WebSocketUpgrade` and `Request<Body>`, I found trait `FromRequest` but don't know how to use it ^^' (sorry I'm fairly noob to Rust).

I have something like :

```rust
async fn matcher(Extension(state): Extension<Arc<State>>, req: TYPE_HERE?) -> impl IntoResponse {
  // ...
  match some_key {
     ws_key => websocket_handler(state, req).await,
      _ => handler(state, req).await
  }
}
```
Handlers :

```rust
// rest
pub async fn handler(state: Arc<State>, mut req: Request<Body>) -> impl IntoResponse
// ws
pub async fn websocket_handler(state: Arc<State>, ws: WebSocketUpgrade) -> impl IntoResponse
```

**Reply by davidpdrsn** [reply]
You have to exact everything you need in the top level function and pass them in to the lower level functions as necessary:

```rust
fn matcher(Extension(state): Extension<Arc<State>>, ws: WebSocketUpgrade, request: Request<Body>)
```

Note that `Request<Body>` must be the last extractor as it consumes the whole request as described here https://docs.rs/axum/latest/axum/extract/index.html#applying-multiple-extractors

**Reply by L4ngu0r** [reply]
Aah yeah! Thanks :pray: sorry I didn't catch this part :disappointed:

**Reply by L4ngu0r** [reply]
Something strange: simple request (REST) failed on 400 because : `Connection header did not include 'upgrade'` if I remove `ws: WebScoketUpgrade` it works :thinking:

**Reply by davidpdrsn** [reply]
https://docs.rs/axum/latest/axum/extract/index.html#optional-extractors
